{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%cd scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-13T06:40:26.417400Z",
     "iopub.status.busy": "2023-07-13T06:40:26.417242Z",
     "iopub.status.idle": "2023-07-13T06:40:26.428582Z",
     "shell.execute_reply": "2023-07-13T06:40:26.427990Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "%matplotlib notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-13T06:40:26.430818Z",
     "iopub.status.busy": "2023-07-13T06:40:26.430522Z",
     "iopub.status.idle": "2023-07-13T06:40:27.436429Z",
     "shell.execute_reply": "2023-07-13T06:40:27.435960Z"
    }
   },
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "import os\n",
    "\n",
    "import graph_tool.all as gt\n",
    "import matplotlib\n",
    "from matplotlib.colors import LogNorm\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import hypergeom, pearsonr\n",
    "import seaborn as sns\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "from functions import *\n",
    "\n",
    "\n",
    "# Graph-Tool compatibility\n",
    "plt.switch_backend('cairo')\n",
    "\n",
    "# Style\n",
    "sns.set_theme(context='talk', style='white', palette='Set2')\n",
    "plt.rcParams.update({\n",
    "    'pdf.fonttype': 42,\n",
    "    'ps.fonttype': 42,\n",
    "    'font.size': 22,\n",
    "    'axes.titlesize': 'medium',\n",
    "    'axes.labelsize': 'large',\n",
    "    'xtick.labelsize': 'medium',\n",
    "    'ytick.labelsize': 'medium',\n",
    "    'legend.fontsize': 'medium',\n",
    "    'legend.title_fontsize': 'medium',\n",
    "    'figure.titlesize': 'x-large',\n",
    "})\n",
    "\n",
    "# Figure transparency\n",
    "# matplotlib.rcParams['figure.facecolor'] = (1., 0., 0., 0.3)  # Debugging\n",
    "matplotlib.rcParams['figure.facecolor'] = (1., 0., 0., 0.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████| 1022/1022 [00:15<00:00, 64.77it/s]\n"
     ]
    }
   ],
   "source": [
    "# Integrity check\n",
    "check_ct_edge_specificity()  # Check for duplicate edges with different attentions"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata\n",
    "meta = get_meta()\n",
    "\n",
    "# Subject preview\n",
    "filtered = []\n",
    "for i, row in meta.iterrows():\n",
    "    try:\n",
    "        load_graph_by_id(row['SubID'])\n",
    "        assert not np.isnan(row['nps_MoodDysCurValue'])  # Has NPS information available\n",
    "        assert row['BRAAK_AD'] in (6,) and row['CERAD'] in (4,) and row['CDRScore'] in (3,)\n",
    "    except:\n",
    "        continue\n",
    "    filtered.append(f'{row[\"SubID\"]} {row[\"Ethnicity\"]} {row[\"Sex\"]}, {row[\"Age\"]}, BRAAK {row[\"BRAAK_AD\"]}, CERAD {row[\"CERAD\"]}, CDR {row[\"CDRScore\"]}, {row[\"Dx\"]}')\n",
    "filtered = np.sort(filtered)\n",
    "for i in range(len(filtered)):\n",
    "    # print(filtered[i])\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Attention Stack"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = './attentions.pkl'\n",
    "if os.path.isfile(fname):\n",
    "    # Load data\n",
    "    with open('./attentions.pkl', 'rb') as f:\n",
    "        all_data = pickle.load(f)\n",
    "    attention_stack, all_edges, columns, subject_ids = all_data['data'], all_data['edges'], all_data['heads'], all_data['subject_ids']\n",
    "\n",
    "else:\n",
    "    # Parameters\n",
    "    # Scaled probably shouldn't be used, but better for visualization\n",
    "    # until results are more even\n",
    "    columns = get_attention_columns(scaled=False)\n",
    "    subject_ids = meta['SubID'].to_numpy()\n",
    "\n",
    "    # Load graphs\n",
    "    graphs, subject_ids = load_many_graphs(subject_ids, column=columns)\n",
    "    # graphs = [compute_graph(g) for g in graphs]\n",
    "\n",
    "    # # Get attentions\n",
    "    # df = {}\n",
    "    # for column in get_attention_columns():\n",
    "    #     attention, _ = compute_edge_summary(graphs, subject_ids=subject_ids)\n",
    "    #     attention = attention.set_index('Edge')\n",
    "    #     df[column] = attention.var(axis=1)\n",
    "\n",
    "\n",
    "    # Set indices to edges and clean\n",
    "    print('Fixing indices...')\n",
    "    for i in tqdm(range(len(graphs))):\n",
    "        graphs[i].index = graphs[i].apply(lambda r: get_edge_string([r['TF'], r['TG']]), axis=1)\n",
    "        graphs[i] = graphs[i].drop(columns=['TF', 'TG'])\n",
    "        # Remove duplicates\n",
    "        graphs[i] = graphs[i][~graphs[i].index.duplicated(keep='first')]\n",
    "\n",
    "    # Get all unique edges\n",
    "    print('Getting unique edges...')\n",
    "    all_edges = np.unique(sum([list(g.index) for g in graphs], []))\n",
    "\n",
    "\n",
    "    # Standardize index order\n",
    "    print('Standardizing indices...')\n",
    "    for i in tqdm(range(len(graphs))):\n",
    "        # Add missing indices and order based on `all_edges`\n",
    "        # to_add = [edge for edge in all_edges if edge not in list(graphs[i].index)]  # SLOW\n",
    "        to_add = list(set(all_edges) - set(graphs[i].index))\n",
    "\n",
    "        # Empty rows\n",
    "        new_rows = pd.DataFrame(\n",
    "            [[np.nan]*len(graphs[i].columns)]*len(to_add),\n",
    "            columns=graphs[i].columns,\n",
    "        ).set_index(pd.Series(to_add))\n",
    "        # Native concat\n",
    "        graphs[i] = pd.concat([graphs[i], new_rows]).loc[all_edges]\n",
    "\n",
    "    # Convert to numpy\n",
    "    graphs = [g.to_numpy() for g in graphs]\n",
    "    attention_stack = np.stack(graphs, axis=-1)\n",
    "    # attention_stack.shape = (Edge, Head, Subject)\n",
    "    # attention_stack.shape = (all_edges, columns, subject_ids)\n",
    "\n",
    "    # Save all data\n",
    "    all_data = {'data': attention_stack, 'edges': all_edges, 'heads': columns, 'subject_ids': subject_ids}\n",
    "    # np.savez('attentions.npz', **all_data)\n",
    "    with open(fname, 'wb') as f:\n",
    "        pickle.dump(\n",
    "            all_data,\n",
    "            f,\n",
    "            protocol=pickle.HIGHEST_PROTOCOL,\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Additional useful parameters\n",
    "self_loops = [split_edge_string(s)[0] == split_edge_string(s)[1] for s in all_edges]\n",
    "self_loops = np.array(self_loops)\n",
    "# Remove self loops\n",
    "all_edges = all_edges[~self_loops]\n",
    "attention_stack = attention_stack[~self_loops]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export example data\n",
    "# for sid_idx in range(5):\n",
    "#     sid = subject_ids[sid_idx]\n",
    "#     data = attention_stack[:, :, sid_idx]\n",
    "#     df = pd.DataFrame(data, index=all_edges, columns=columns).dropna()\n",
    "#     df.to_csv(f'{sid}.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Global Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Available attention columns: ['AD_imp_1', 'AD_imp_2', 'SCZ_imp_1', 'SCZ_imp_2', 'data_imp_1', 'data_imp_2', 'data_imp_3', 'data_imp_4']\n"
     ]
    }
   ],
   "source": [
    "# Parameters\n",
    "print(f'\\nAvailable attention columns: {get_attention_columns()}')\n",
    "column_ad = get_attention_columns()[0]\n",
    "column_scz = get_attention_columns()[2]\n",
    "column_data = get_attention_columns()[4]\n",
    "synthetic_nodes_of_interest = ['OPC', 'Micro', 'Oligo']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intra-Contrast Comparisons (Figure 4/6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure parameters\n",
    "param = {\n",
    "    'subjects': ['M31969', 'M20337'],\n",
    "    'columns': [column_data, column_ad, column_scz],\n",
    "    'column_names': ['Data-Driven', 'AD-Prior', 'SCZ-Prior'],\n",
    "    'column_groups': [get_attention_columns()[4:8], get_attention_columns()[:2], get_attention_columns()[2:4]],\n",
    "    'column_group_names': ['Data Prioritization', 'AD Prioritization', 'SCZ Prioritization'],\n",
    "    'ancestries': meta.groupby('Ethnicity').count()['SubID'].sort_values().index[::-1].to_list()[:3] + ['all'],\n",
    "    'contrast': 'c15x',\n",
    "}\n",
    "\n",
    "# Generate palette\n",
    "palette = plt.rcParams['axes.prop_cycle'].by_key()['color']\n",
    "param['palette'] = {sid: rgba_to_hex(palette[i]) for i, sid in enumerate(param['subjects'])}\n",
    "\n",
    "# Preview subjects\n",
    "for sid in param['subjects']:\n",
    "    row = meta.loc[meta['SubID']==sid].iloc[0]\n",
    "    # print(f'{row[\"SubID\"]} {row[\"Ethnicity\"]} {row[\"Sex\"]}, {row[\"Age\"]}, BRAAK {row[\"BRAAK_AD\"]}, CERAD {row[\"CERAD\"]}, CDR {row[\"CDRScore\"]}, {row[\"Dx\"]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subplot layout (doesn't work well with constrained layout)\n",
    "# NOTE: This cannot be used, as constrained layout has glitches\n",
    "# (see https://github.com/matplotlib/matplotlib/issues/23290)\n",
    "# with uneven mosaics\n",
    "# fig, axs = get_mosaic(shape, figsize=(int((3/2) * shape_array.shape[1]), int((3/2) * shape_array.shape[0])), constrained_layout=False)\n",
    "\n",
    "# Subfigure layout (longer)\n",
    "# NOTE: Constrained layout will fail for all\n",
    "# subplots if a single one is not able to scale.\n",
    "# Also, sometimes leaving a subfigure blank will\n",
    "# cause it to fail, especially if on an edge.\n",
    "# It is VERY finnicky.\n",
    "# SOLUTION: Save again using `fig.savefig(...)`\n",
    "# and it will run without warning.  Then, you\n",
    "# can visually inspect for scaling issues.\n",
    "# fig, axs = create_subfigure_mosaic(shape_array)\n",
    "# fig.set_constrained_layout_pads(w_pad=0, h_pad=0, wspace=.4, hspace=.4)  # *_pad is pad for figs (including subfigs), *_space is pad between subplots\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.unique(meta.loc[meta['SubID'].isin(all_data['subject_ids']), 'Ethnicity'], return_counts=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main (Figure 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Edge Discovery Enrichment (None, None, None, N)\n",
      "Prioritization Ranges [[14.0, 22.0], [22.0, 49.0], [49.0, 680.0]]\n",
      "Filtered 165669 edges of 204507 total from histogram\n",
      "Prioritization Ranges [[9.0, 13.0], [13.0, 25.0], [25.0, 104.0]]\n",
      "Filtered 52826 edges of 61429 total from histogram\n",
      "Prioritization Ranges [[8.0, 11.0], [11.0, 21.0], [21.0, 88.0]]\n",
      "Filtered 46979 edges of 53899 total from histogram\n",
      "Prioritization Ranges [[15.0, 23.0], [23.0, 51.0], [51.0, 875.0]]\n",
      "Filtered 195859 edges of 242486 total from histogram\n",
      "\n",
      "Ancestry Enrichment Comparison (R)\n",
      "Index(['AFR_data_imp_1', 'AMR_data_imp_1', 'EUR_data_imp_1', 'all_data_imp_1'], dtype='object', name='Ancestry')\n",
      "Index(['EUR', 'AFR', 'AMR', 'all'], dtype='object')\n",
      "Index(['EUR', 'AFR', 'AMR'], dtype='object')\n",
      "\n",
      "Saving Figure...\n"
     ]
    }
   ],
   "source": [
    "# Top part\n",
    "# shape = \"\"\"\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBB...........\n",
    "#     .AAAAAABBBBBBRRRRRRRRRRR\n",
    "#     .AAAAAABBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "# \"\"\"\n",
    "shape = \"\"\"\n",
    "    NNNNNNNNNNN\n",
    "    NNNNNNNNNNN\n",
    "    NNNNNNNNNNN\n",
    "    NNNNNNNNNNN\n",
    "    RRRRRRRRRRR\n",
    "    RRRRRRRRRRR\n",
    "    RRRRRRRRRRR\n",
    "    RRRRRRRRRRR\n",
    "    RRRRRRRRRRR\n",
    "    RRRRRRRRRRR\n",
    "    RRRRRRRRRRR\n",
    "\"\"\"\n",
    "fig, axs = create_subfigure_mosaic(shape_array_from_shape(shape))\n",
    "\n",
    "# Plot all panels\n",
    "# axs_lab = ['A', 'G']\n",
    "# print(f'Individual Panels ({\", \".join(axs_lab)})')\n",
    "# plot_graph_comparison_from_sids(param['subjects'], axs={0: axs[axs_lab[0]], 1: axs[axs_lab[1]]}, column=param['columns'][0], vertex_ids=synthetic_nodes_of_interest)\n",
    "# # Legend\n",
    "# plot_legend(horizontal=False, loc='center', bbox_to_anchor=(.5, -.1), ax=axs[axs_lab[0]])\n",
    "\n",
    "# axs_lab = ['B']\n",
    "# print(f'Data Heatmap ({\", \".join(axs_lab)})')\n",
    "# ax=axs[axs_lab[0]] = plot_heatmap_comparison(\n",
    "#     **all_data,\n",
    "#     target_sids=param['subjects'],\n",
    "#     ax=axs[axs_lab[0]],\n",
    "#     column_groups=param['column_groups'],\n",
    "#     column_group_names=param['column_group_names'],\n",
    "#     palette=param['palette'],\n",
    "#     random_seed=42,\n",
    "#     num_edges=40)\n",
    "\n",
    "axs_lab = (len(param['ancestries']) - 1) * ['None'] + ['N']\n",
    "# axs_lab = ['K', 'L', 'M', 'N']\n",
    "print(f'\\nEdge Discovery Enrichment ({\", \".join(axs_lab)})')\n",
    "for ancestry, ax in zip(param['ancestries'], [axs[lab] if lab in axs else None for lab in axs_lab]):\n",
    "    # Filter to ancestry\n",
    "    anc_data = all_data.copy()\n",
    "    if ancestry != 'all':\n",
    "        sub_ids = meta.loc[meta['Ethnicity'] == ancestry, 'SubID'].to_list()\n",
    "        mask = [sid in sub_ids for sid in anc_data['subject_ids']]\n",
    "        anc_data['data'] = anc_data['data'][:, :, mask]\n",
    "        anc_data['subject_ids'] = np.array(anc_data['subject_ids'])[mask]\n",
    "\n",
    "    # Run\n",
    "    temp = plot_edge_discovery_enrichment(\n",
    "        **anc_data,\n",
    "        column=param['columns'][0],\n",
    "        range_colors=[rgb_to_float(hex_to_rgb('#7aa457')), rgb_to_float(hex_to_rgb('#a46cb7')), rgb_to_float(hex_to_rgb('#cb6a49'))],\n",
    "        ax=ax,\n",
    "        postfix=f'{ancestry}_{param[\"columns\"][0]}',\n",
    "        gene_max_num=300,\n",
    "        threshold=95,\n",
    "        clamp_min=4,\n",
    "        skip_plot=(ax is None),\n",
    "        verbose=True)\n",
    "    if ax is not None:\n",
    "        ax.set_xlabel(f'High-Scoring Edges ({param[\"column_names\"][0]})')\n",
    "        ylabel = 'Frequency'\n",
    "        if ancestry != 'all': ylabel += f' ({ancestry})'\n",
    "        ax.set_ylabel(ylabel)\n",
    "    # MANUAL PROCESSING\n",
    "    # Run the output '../plots/genes_<column>.csv' from above on Metascape as multiple gene list and perform\n",
    "    # enrichment.  From the all-in-one ZIP file, save the file from Enrichment_GO/GO_membership.csv as '../plots/go_<column>.csv'\n",
    "    # and rerun.\n",
    "\n",
    "axs_lab = ['R']\n",
    "print(f'\\nAncestry Enrichment Comparison ({\", \".join(axs_lab)})')\n",
    "postfixes = [f'{ancestry}_{param[\"columns\"][0]}' for ancestry in param['ancestries']]\n",
    "enrichments = plot_cross_enrichment(postfixes, names=param['ancestries'], ax=axs[axs_lab[0]], excluded_subgroups=['all'])\n",
    "\n",
    "# Place labels\n",
    "offset = plot_labels(axs, shape=shape)\n",
    "\n",
    "# Save figure\n",
    "print('\\nSaving Figure...')\n",
    "fig.savefig(f'../plots/figure_4_main.pdf', bbox_inches='tight', pad_inches=1, format='pdf', transparent=True, backend='cairo')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplement (Figure 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Edge Discovery Enrichment (K, L, M, N, O, P, Q, R, S, T, U, V)\n",
      "\n",
      "Saving Figure...\n"
     ]
    }
   ],
   "source": [
    "# Top part\n",
    "# shape = \"\"\"\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBBNNNNNNNNNNN\n",
    "#     .AAAAAABBBBBB...........\n",
    "#     .AAAAAABBBBBBRRRRRRRRRRR\n",
    "#     .AAAAAABBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "#     .GGGGGGBBBBBBRRRRRRRRRRR\n",
    "# \"\"\"\n",
    "shape = \"\"\"\n",
    "    KKKKKKKKKKKLLLLLLLLLLLMMMMMMMMMMM\n",
    "    KKKKKKKKKKKLLLLLLLLLLLMMMMMMMMMMM\n",
    "    KKKKKKKKKKKLLLLLLLLLLLMMMMMMMMMMM\n",
    "    KKKKKKKKKKKLLLLLLLLLLLMMMMMMMMMMM\n",
    "    NNNNNNNNNNNOOOOOOOOOOOPPPPPPPPPPP\n",
    "    NNNNNNNNNNNOOOOOOOOOOOPPPPPPPPPPP\n",
    "    NNNNNNNNNNNOOOOOOOOOOOPPPPPPPPPPP\n",
    "    NNNNNNNNNNNOOOOOOOOOOOPPPPPPPPPPP\n",
    "    QQQQQQQQQQQRRRRRRRRRRRSSSSSSSSSSS\n",
    "    QQQQQQQQQQQRRRRRRRRRRRSSSSSSSSSSS\n",
    "    QQQQQQQQQQQRRRRRRRRRRRSSSSSSSSSSS\n",
    "    QQQQQQQQQQQRRRRRRRRRRRSSSSSSSSSSS\n",
    "    TTTTTTTTTTTUUUUUUUUUUUVVVVVVVVVVV\n",
    "    TTTTTTTTTTTUUUUUUUUUUUVVVVVVVVVVV\n",
    "    TTTTTTTTTTTUUUUUUUUUUUVVVVVVVVVVV\n",
    "    TTTTTTTTTTTUUUUUUUUUUUVVVVVVVVVVV\n",
    "\"\"\"\n",
    "fig, axs = create_subfigure_mosaic(shape_array_from_shape(shape))\n",
    "\n",
    "# Plot all panels\n",
    "axs_lab = ['K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V']\n",
    "print(f'\\nEdge Discovery Enrichment ({\", \".join(axs_lab)})')\n",
    "for (ancestry, column_idx), ax in zip(product(param['ancestries'], range(len(param['columns']))), [axs[lab] if lab in axs else None for lab in axs_lab]):\n",
    "    # Filter to ancestry\n",
    "    anc_data = all_data.copy()\n",
    "    if ancestry != 'all':\n",
    "        sub_ids = meta.loc[meta['Ethnicity'] == ancestry, 'SubID'].to_list()\n",
    "        mask = [sid in sub_ids for sid in anc_data['subject_ids']]\n",
    "        anc_data['data'] = anc_data['data'][:, :, mask]\n",
    "        anc_data['subject_ids'] = np.array(anc_data['subject_ids'])[mask]\n",
    "\n",
    "    # Run\n",
    "    temp = plot_edge_discovery_enrichment(\n",
    "        **anc_data,\n",
    "        column=param['columns'][column_idx],\n",
    "        range_colors=[rgb_to_float(hex_to_rgb('#7aa457')), rgb_to_float(hex_to_rgb('#a46cb7')), rgb_to_float(hex_to_rgb('#cb6a49'))],\n",
    "        ax=ax,\n",
    "        postfix=f'{ancestry}_{param[\"columns\"][column_idx]}',\n",
    "        gene_max_num=300,\n",
    "        threshold=95,\n",
    "        skip_plot=(ax is None))\n",
    "    if ax is not None:\n",
    "        ax.set_xlabel(f'High-Scoring Edges ({param[\"column_names\"][column_idx]})')\n",
    "        ylabel = 'Frequency'\n",
    "        if ancestry != 'all': ylabel += f' ({ancestry})'\n",
    "        ax.set_ylabel(ylabel)\n",
    "    # MANUAL PROCESSING\n",
    "    # Run the output '../plots/genes_<column>.csv' from above on Metascape as multiple gene list and perform\n",
    "    # enrichment.  From the all-in-one ZIP file, save the file from Enrichment_GO/GO_membership.csv as '../plots/go_<column>.csv'\n",
    "    # and rerun.\n",
    "\n",
    "# Place labels\n",
    "offset = plot_labels(axs, shape=shape)\n",
    "\n",
    "# Save figure\n",
    "print('\\nSaving Figure...')\n",
    "fig.savefig(f'../plots/figure_4_supplement.pdf', bbox_inches='tight', pad_inches=1, format='pdf', transparent=True, backend='cairo')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PRS Analyses (Figure 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PRS Analysis (S, U)\n",
      "\n",
      "Saving Figure...\n"
     ]
    }
   ],
   "source": [
    "# SCZ\n",
    "# Data\n",
    "# AD\n",
    "# SCZ\n",
    "shape = \"\"\"\n",
    "    SSSSSSSSSSSSSSSUUUUUUUUUUUUUUU\n",
    "    SSSSSSSSSSSSSSSUUUUUUUUUUUUUUU\n",
    "    SSSSSSSSSSSSSSSUUUUUUUUUUUUUUU\n",
    "    SSSSSSSSSSSSSSSUUUUUUUUUUUUUUU\n",
    "\"\"\"\n",
    "fig, axs = create_subfigure_mosaic(shape_array_from_shape(shape))\n",
    "\n",
    "# Plot all panels\n",
    "axs_lab = ['S', 'U']\n",
    "print(f'\\nPRS Analysis ({\", \".join(axs_lab)})')\n",
    "# Takes around an hour for each loop with no subsampling (on first run)\n",
    "for fname, head_prefix, ylabel, prs_col, ax_idx in zip(\n",
    "    ('ad_prs_df.csv', 'scz_prs_df.csv'),\n",
    "    ('_'.join(column_ad.split('_')[:-1]), '_'.join(column_scz.split('_')[:-1])),\n",
    "    ('AD Importance Score', 'SCZ Importance Score'),\n",
    "    ('prs_scaled_AD_Bellenguez', 'prs_scaled_SCZ.3.5_MVP'),\n",
    "    axs_lab\n",
    "):\n",
    "    df = pd.read_csv(fname, index_col=0) if os.path.isfile(fname) else None\n",
    "    covariates = get_genotype_meta()[['SubID', 'imp_sex_score'] + [f'imp_anc_PC{i}' for i in range(1, 7)] + [f'imp_anc_{anc}' for anc in ('AFR', 'AMR', 'EAS', 'EUR')]]\n",
    "    df, prs_df, axs[ax_idx] = plot_prs_correlation(\n",
    "        meta, **all_data, ax=axs[ax_idx],\n",
    "        df=df, num_targets=5, ylabel=ylabel, max_scale=False,\n",
    "        head_prefix=head_prefix, prs_col=prs_col,\n",
    "        covariates=covariates, subsample=1)\n",
    "    if not os.path.isfile(fname): df.to_csv(fname)\n",
    "\n",
    "# Place labels\n",
    "offset = plot_labels(axs, shape=shape)\n",
    "\n",
    "# Save figure\n",
    "print('\\nSaving Figure...')\n",
    "fig.savefig(f'../plots/figure_6_prs.pdf', bbox_inches='tight', pad_inches=1, format='pdf', transparent=True, backend='cairo')\n",
    "\n",
    "# CLI\n",
    "# df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extended (Figure 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Extended\n",
    "# shape = \"\"\"\n",
    "#     EEEEEEFFFFFF\n",
    "#     EEEEEEFFFFFF\n",
    "#     EEEEEEFFFFFF\n",
    "#     BBBBCCCCDDDD\n",
    "#     BBBBCCCCDDDD\n",
    "#     BBBBCCCCDDDD\n",
    "#     BBBBCCCCDDDD\n",
    "#     BBBBCCCCDDDD\n",
    "#     HHHHIIIIJJJJ\n",
    "#     HHHHIIIIJJJJ\n",
    "#     HHHHIIIIJJJJ\n",
    "# \"\"\"\n",
    "# fig, axs = create_subfigure_mosaic(shape_array_from_shape(shape))\n",
    "\n",
    "# # Plot all panels\n",
    "# axs_lab = ['B', 'D']\n",
    "# print(f'\\nCell Type Graph ({\", \".join(axs_lab)})')\n",
    "# pos_dict = {}\n",
    "# for ax, column, title in zip([axs[lab] for lab in axs_lab], param['columns'], param['column_names']):\n",
    "#     pos_dict = plot_ct_graph_from_sid(param['subjects'][0], ax=ax, column=column, **pos_dict)\n",
    "#     ax.set_title(title)\n",
    "\n",
    "# axs_lab = ['C']\n",
    "# print(f'\\nCell Type Priority ({\", \".join(axs_lab)})')\n",
    "# plot_ct_individual_edge_comparison(param['subjects'][0], **all_data, columns=param['columns'][:2], column_names=param['column_names'][:2], ax=axs[axs_lab[0]])\n",
    "# axs[axs_lab[0]].set_xlabel(None)\n",
    "\n",
    "# axs_lab = ['E', 'F']\n",
    "# print(f'\\nAttention Histogram ({\", \".join(axs_lab)})')\n",
    "# ax_ymax = 1\n",
    "# for ax, column, title in zip([axs[lab] for lab in axs_lab], param['columns'], param['column_names']):\n",
    "#     plot_attention_histogram(param['subjects'][0], **all_data, column=column, ax=ax)\n",
    "#     ax.set_xlabel(title)\n",
    "#     # Record limits\n",
    "#     ylim = ax.get_ylim()\n",
    "#     if ylim[1] > ax_ymax: ax_ymax = ylim[1]\n",
    "# # Set axis limits\n",
    "# for ax in [axs[lab] for lab in axs_lab]:\n",
    "#     ax.set_ylim(1, ax_ymax)\n",
    "\n",
    "# axs_lab = ['H', 'I', 'J']\n",
    "# print(f'\\nEdge Comparisons ({\", \".join(axs_lab)})')\n",
    "# for ax, column, title in zip([axs[lab] for lab in axs_lab], param['column_groups'], param['column_group_names']):\n",
    "#     ax.clear()\n",
    "#     plot_edge_comparison_from_sids(param['subjects'], ax=ax, column=column, palette=param['palette'], highlight_outliers=True)\n",
    "#     ax.set_title(title)\n",
    "\n",
    "# # Place labels\n",
    "# offset = plot_labels(axs, shape=shape)\n",
    "\n",
    "# # Save figure\n",
    "# print('\\nSaving Figure...')\n",
    "# fig.savefig(f'../plots/figure_4_extended.pdf', bbox_inches='tight', pad_inches=1, format='pdf', transparent=True, backend='cairo')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Unused\n",
    "# shape = \"\"\"\n",
    "#     LLLLLLLMMMMMMMNNNNNNN\n",
    "#     LLLLLLLMMMMMMMNNNNNNN\n",
    "#     LLLLLLLMMMMMMMNNNNNNN\n",
    "#     LLLLLLLMMMMMMMNNNNNNN\n",
    "#     LLLLLLLMMMMMMMNNNNNNN\n",
    "# \"\"\"\n",
    "# fig, axs = create_subfigure_mosaic(shape_array_from_shape(shape))\n",
    "\n",
    "# # Plot all panels\n",
    "# axs_lab = ['K']\n",
    "# print(f'\\nModule Analysis ({\", \".join(axs_lab)})')\n",
    "# plot_module_scores_from_sids(param['subjects'], ax=axs[axs_lab[0]], palette=param['palette'], column=param['columns'][0])\n",
    "# axs[axs_lab[0]].set_title(param['column_names'][0])\n",
    "\n",
    "# axs_lab = ['L']\n",
    "# print(f'\\nOutcome Icon ({\", \".join(axs_lab)})')\n",
    "# # Manual, so no visualization is needed\n",
    "# axs[axs_lab[0]].axis('off')\n",
    "\n",
    "# axs_lab = ['M']\n",
    "# print(f'\\nEdge Discovery ({\", \".join(axs_lab)})')\n",
    "# plot_edge_discovery(**all_data, column=param['columns'][0], ax=axs[axs_lab[0]])\n",
    "# axs[axs_lab[0]].set_title(param['column_names'][0])\n",
    "\n",
    "# axs_lab = ['N']\n",
    "# print(f'\\nEnrichment ({\", \".join(axs_lab)})')\n",
    "# # Compute gene list\n",
    "# fname = f'../plots/genes_{\"_\".join(param[\"subjects\"])}.csv'\n",
    "# if not os.path.isfile(fname):\n",
    "#     df = compute_all_important_genes_from_sids(param['subjects'], **all_data, vertex_ids=synthetic_nodes_of_interest, columns=param['columns'], column_names=param['column_names'])\n",
    "#     df.to_csv(fname, index=False)\n",
    "# # MANUAL PROCESSING\n",
    "# # Run the output from above on Metascape as multiple gene list and perform\n",
    "# # enrichment.  From the all-in-one ZIP file, save the file from\n",
    "# # Enrichment_GO/GO_membership.csv as `fname` below\n",
    "# # Read enrichment\n",
    "# fname = f'../plots/go_{\"_\".join(param[\"subjects\"])}.csv'\n",
    "# if os.path.isfile(fname):\n",
    "#     plot_enrichment_from_fname(fname, ax=axs[axs_lab[0]])\n",
    "# axs[axs_lab[0]].set_xlabel(None)\n",
    "\n",
    "# # Inset axes\n",
    "# from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
    "# axins = inset_axes(\n",
    "#     axs[axs_lab[0]],\n",
    "#     width='20%', height='50%',\n",
    "#     loc=1,\n",
    "#     bbox_to_anchor=(-1, -1.3, 1, 1), bbox_transform=axs[axs_lab[0]].transAxes)\n",
    "\n",
    "# # axs_lab = ['M']\n",
    "# print(f'\\nEdge Discovery ({\", \".join(axs_lab)})')\n",
    "# plot_edge_discovery(**all_data, column=param['columns'][0], ax=axins)  # axs[axs_lab[0]]\n",
    "# axins.set_title(param['column_names'][0])\n",
    "# axins.set_xlabel(None)\n",
    "\n",
    "# # Place labels\n",
    "# # offset = plot_labels(axs, shape=shape)\n",
    "\n",
    "# # Save figure\n",
    "# print('\\nSaving Figure...')\n",
    "# fig.savefig(f'../plots/figure_3_unused.pdf', bbox_inches='tight', pad_inches=.5, format='pdf', transparent=True, backend='cairo')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inter-Contrast Comparisons (Figure 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Combinations\n",
    "# # TODO: Potentially move each entry to dictionary, so changes in order\n",
    "# #   are easier to propagate\n",
    "# contrast_groupings = [\n",
    "#     # (contrast name, contrast group, attention column, comparison column, target meta column, other target meta column)\n",
    "#     # for contrast_name, contrast_group, column, comparison, target, target_comparison in contrast_groupings:\n",
    "#     # TODO: Revise ethnicity prediction\n",
    "#     ('c15x', 'AD', column_ad, column_data, 'BRAAK_AD', 'Ethnicity'),\n",
    "#     ('c15x', 'AD', column_data, column_ad, 'BRAAK_AD', 'Ethnicity'),\n",
    "#     ('c15x', 'AD', column_data, column_ad, 'nps_MoodDysCurValue', 'nps_WtGainCurValue'),\n",
    "#     # ('c06x', 'AD', column_ad, column_data, 'BRAAK_AD', 'nps_MoodDysCurValue'),  # Eventually SCZ, BP and such\n",
    "#     # ('c71x', 'MoodDys', column_data, column_ad, 'nps_MoodDysCurValue'),  # Dysphoria\n",
    "#     # ('c72x', 'DecInt', column_data, column_ad, 'nps_DecIntCurValue'),  # Anhedonia\n",
    "# ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4B Distribution Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for contrast_name, _, column, comparison, target, target_comparison in contrast_groupings:\n",
    "#     # Filter attention stack to contrast\n",
    "#     contrast = get_contrast(contrast_name)\n",
    "#     contrast_subject_ids = sum([contrast[group] for group in contrast], [])\n",
    "#     contrast_mask = [sid in contrast_subject_ids for sid in subject_ids]\n",
    "#     contrast_subject_ids = np.array(subject_ids)[contrast_mask]\n",
    "#     contrast_stack = attention_stack[:, :, contrast_mask]\n",
    "\n",
    "#     # Filter to 1000 most variant edges\n",
    "#     top_variant_edge_idx = np.nan_to_num(\n",
    "#         contrast_stack[:, np.argwhere(np.array(columns)==column)[0][0]]).var(axis=1).argsort()[::-1][:1000]\n",
    "#     contrast_stack = contrast_stack[top_variant_edge_idx]\n",
    "#     edge_names = all_edges[top_variant_edge_idx]\n",
    "\n",
    "#     # Correlation df\n",
    "#     df = pd.DataFrame(\n",
    "#         contrast_stack[:, np.argwhere(np.array(columns)==column)[0][0]],\n",
    "#         index=pd.Series(all_edges[top_variant_edge_idx]),\n",
    "#         columns=contrast_subject_ids).T\n",
    "#     df = df.join(meta.set_index('SubID')[[target, target_comparison]]).reset_index(drop=True)\n",
    "#     # Select edge which most cleanly separates `target`\n",
    "#     # top_distinct_edge_idx = df.drop(target_comparison, axis=1).groupby(target).mean().var(axis=0).argsort()[-1]\n",
    "#     # Select top 3 most correlating edges\n",
    "#     edge_name = df.drop(target_comparison, axis=1).corr()[target].abs().drop(target).sort_values(ascending=False)[:3].index.to_numpy()\n",
    "#     top_distinct_edge_idx = [np.argwhere(df.columns==edge)[0][0] for edge in edge_name]\n",
    "#     contrast_stack = contrast_stack[top_distinct_edge_idx]\n",
    "\n",
    "#     # Plot\n",
    "#     fig, axs = get_mosaic(np.array(sum([[i]*2 for i in range(6)], [])).reshape((2, -1)), scale=5)\n",
    "#     # axs[0].sharex(axs[1])\n",
    "#     sns.despine()\n",
    "\n",
    "#     for i in range(3):\n",
    "#         # Filter\n",
    "#         contrast_stack_i = contrast_stack[i]\n",
    "#         edge_name_i = edge_name[i]\n",
    "\n",
    "#         # Scale attention\n",
    "#         # TODO: Remove once heads are balanced\n",
    "#         contrast_stack_i = contrast_stack_i / np.nan_to_num(contrast_stack_i).max(axis=1).reshape((-1, 1))\n",
    "\n",
    "#         # Format\n",
    "#         df = pd.DataFrame(contrast_stack_i, index=pd.Series(columns), columns=contrast_subject_ids)\n",
    "#         df = df.reset_index(names='Head').melt(id_vars='Head', var_name='Subject', value_name=edge_name_i).dropna()  # Melt\n",
    "#         df = df.set_index('Subject').join(meta.set_index('SubID')[[target, target_comparison]]).reset_index()  # Join meta\n",
    "\n",
    "#         # Filter to target heads\n",
    "#         df = df.loc[df['Head'].apply(lambda s: s in (column, comparison))]\n",
    "\n",
    "#         # Main target\n",
    "#         p1 = sns.violinplot(data=df, x='Head', y=edge_name_i, hue=target, ax=axs[i])\n",
    "#         p1.legend(title=target, bbox_to_anchor=(1.1, 1.05))\n",
    "#         p1.set(xlabel=None, xticklabels=[])\n",
    "\n",
    "#         # Comparison target\n",
    "#         p2 = sns.violinplot(data=df, x='Head', y=edge_name_i, hue=target_comparison, ax=axs[i+3])\n",
    "#         p2.legend(title=target_comparison, bbox_to_anchor=(1.1, 1.05))\n",
    "#         plt.sca(p2)\n",
    "#         plt.xticks(rotation=60)\n",
    "\n",
    "#         # Get correlation p-values for targets (which must be numeric)\n",
    "#         for j, tar in enumerate((target, target_comparison)):\n",
    "#             for k, c in enumerate(np.unique(df['Head'])):\n",
    "#                 try:\n",
    "#                     pval = pearsonr(df.loc[df['Head']==c, edge_name_i], df.loc[df['Head']==c, tar])[1]\n",
    "#                     axs[i+3*j].text(k, axs[i+3*j].get_ylim()[0] - (.15 if not j else .3), f'p={pval:.1e}', ha='center', va='center')\n",
    "#                 except: continue\n",
    "\n",
    "#     fig.savefig(f'../plots/group_differential_expression_{contrast_name}_{column}_{comparison}_{target}_{target_comparison}.pdf', format='pdf', transparent=True, backend='cairo')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4CD Linkage Cluster Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for contrast_name, _, column, _, target, target_comparison in contrast_groupings:\n",
    "#     for tar in (target, target_comparison):\n",
    "#         # Get subject ids\n",
    "#         group = None  # contrast_group\n",
    "#         if group is None:\n",
    "#             # Population\n",
    "#             contrast_subjects = sum([v for k, v in get_contrast(contrast_name).items()], [])\n",
    "#         else:\n",
    "#             # Group\n",
    "#             contrast_subjects = get_contrast(contrast_name)[group]\n",
    "\n",
    "#         # Modify stack to include only contrast\n",
    "#         df = np.nan_to_num(attention_stack[:, np.argwhere(np.array(columns)==column)[0][0], [s in contrast_subjects for s in subject_ids]])\n",
    "#         new_subject_ids = [s for s in subject_ids if s in contrast_subjects]\n",
    "#         df = pd.DataFrame(df, index=all_edges, columns=new_subject_ids)\n",
    "\n",
    "#         # Get 100 most variant edges\n",
    "#         df = df.iloc[df.to_numpy().var(axis=1).argsort()[::-1][:100]]\n",
    "\n",
    "#         # Cluster\n",
    "#         labels = KMeans(n_clusters=10, n_init=10).fit_predict(df.to_numpy().T)\n",
    "#         labels += 1\n",
    "\n",
    "#         # Get phenotypes\n",
    "#         pheno = [meta.iloc[np.argwhere(meta['SubID'] == sid)[0][0]][tar] for sid in new_subject_ids]\n",
    "\n",
    "#         # Format results\n",
    "#         df = pd.DataFrame({'Cluster': labels, tar: pheno}, index=new_subject_ids)\n",
    "#         df['count'] = 1\n",
    "#         df = df.pivot_table(index='Cluster', columns=tar, values='count', aggfunc='sum').fillna(0)\n",
    "\n",
    "#         # Transform to hypergeometric\n",
    "#         df_np = df.to_numpy()\n",
    "#         df_np_new = np.zeros_like(df_np)\n",
    "#         for i, j in product(*[range(k) for k in df.shape]):\n",
    "#             # i - cluster, j - target\n",
    "#             dist = hypergeom(df_np.sum(), df_np[:, j].sum(), df_np[i, :].sum())\n",
    "#             # Calculate probability of overrepresentation\n",
    "#             df_np_new[i, j] = 1 - dist.cdf(df_np[i, j])\n",
    "#         with np.errstate(divide='ignore'):\n",
    "#             df_np_new = -np.log10(df_np_new)\n",
    "#             df_np_new[np.isinf(df_np_new)] = np.nan\n",
    "#         df = pd.DataFrame(df_np_new, index=df.index, columns=df.columns)\n",
    "\n",
    "#         # Plot\n",
    "#         fig, axs = get_mosaic([list(range(1))], scale=9)\n",
    "#         sns.heatmap(df, vmin=0, vmax=3, cmap='rocket_r', cbar_kws={'label': '-log10(p)'}, ax=axs[0])\n",
    "#         # plt.tight_layout()\n",
    "#         fig.savefig(f'../plots/group_linkage_cluster_{contrast_name}_{column}_{tar}.pdf', format='pdf', transparent=True, backend='cairo')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4E Aggregate Graph Enrichment (MANUAL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # NOTE: Only top 100 are taken for aggregate due to memory concerns\n",
    "# for contrast_name, group, column, _, _, _ in contrast_groupings:\n",
    "#       # Load contrast\n",
    "#       np.random.seed(42)\n",
    "#       contrast_subjects = get_contrast(contrast_name)\n",
    "#       gs = {\n",
    "#             gname: concatenate_graphs(*[\n",
    "#                   compute_graph(g)\n",
    "#                   for g in load_many_graphs(np.random.choice(sids, 100, replace=False))[0]\n",
    "#             ])\n",
    "#             for gname, sids in contrast_subjects.items()\n",
    "#       }\n",
    "\n",
    "#       # Split into groups\n",
    "#       # TODO: Make more general, perhaps add comparison group to arguments\n",
    "#       g1_name = group\n",
    "#       g1 = gs[g1_name]\n",
    "#       g2_name = 'Control'\n",
    "#       g2 = gs[g2_name]\n",
    "\n",
    "#       # Get unique TFs\n",
    "#       df = compare_graphs_enrichment(\n",
    "#             g1,\n",
    "#             g2,\n",
    "#             sid_1=g1_name,\n",
    "#             sid_2=g2_name,\n",
    "#             nodes=synthetic_nodes_of_interest,\n",
    "#             threshold=.01)\n",
    "\n",
    "#       # Save to file\n",
    "#       df.to_csv(f'../plots/genes_{contrast_name}_{group}_{column}.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Enrichment\n",
    "# for contrast_name, group, column, _, _, _ in contrast_groupings:\n",
    "#     # MANUAL PROCESSING\n",
    "#     # Run the output from above on Metascape as multiple gene list and perform\n",
    "#     # enrichment.  From the all-in-one ZIP file, save the file from\n",
    "#     # Enrichment_QC/GO_DisGeNET as '../plot/disgenet_{subject_id_1}_{subject_id_2}_{column}.csv' and\n",
    "#     # Overlap_circos/CircosOverlapByGene.svg as '../plot/overlap_{subject_id_1}_{subject_id_2}_{column}.svg'\n",
    "\n",
    "#     # Get enrichment\n",
    "#     enrichment_file = f'../plots/disgenet_{contrast_name}_{group}_{column}.csv'\n",
    "#     if enrichment_file is None: continue\n",
    "#     enrichment = pd.read_csv(enrichment_file)\n",
    "\n",
    "#     # Format\n",
    "#     enrichment = format_enrichment(enrichment)\n",
    "\n",
    "#     # Plot\n",
    "#     fig, axs = get_mosaic([[0]*2], scale=9)\n",
    "#     pl = sns.scatterplot(\n",
    "#         enrichment,\n",
    "#         x='Gene Set', y='Description',\n",
    "#         size='-log10(p)',\n",
    "#         color='black',\n",
    "#         ax=axs[0])\n",
    "#     # Formatting\n",
    "#     pl.grid()\n",
    "#     plt.xticks(rotation=90)\n",
    "#     pl.set_aspect('equal', 'box')\n",
    "#     pl.legend(title='-log10(p)', bbox_to_anchor=(1.2, 1.05))\n",
    "#     # Zoom X1\n",
    "#     margin = .5\n",
    "#     min_xlim, max_xlim = pl.get_xlim()\n",
    "#     min_xlim -= margin; max_xlim += margin\n",
    "#     pl.set(xlim=(min_xlim, max_xlim))\n",
    "#     # Save\n",
    "#     fig.savefig(f'../plots/group_enrichment_{contrast_name}_{group}_{column}.pdf', format='pdf', transparent=True, backend='cairo')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
